### PR TITLE
attributed medium and sources on event level

### PR DESCRIPTION
**Pull request by jasoncuts**
Changes made:
1. Cleaned up the existing models with correct column names.
2. Excluded few models from dbt build where the corresponding tables are not ,
3. Added attribution macro to each model wherever we have context_referrer columnto get the marketing_source and marketing_medium.
4. Stitched session to events

### CODE REVIEW

File: models/base/scratch/fueled_events_base_session_info.sql [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+) 
+
+select 
+  * except(landing_page_url),
+  case when landing_page_url like '%checkout%' then 'https://shop.cutsclothing.com/checkouts/'
+  when landing_page_url like '%account%' then 'https://www.cutsclothing.com/account/'
+  else landing_page_url end as landing_page_url
+from final
+qualify row_number() over(partition by session_id order by timestamp asc) = 1
+
```
Comment (farmanp) [inline models/base/scratch/fueled_events_base_session_info.sql:54]: We may need to bring this up as a larger discussion but: 

Typically, we enforce an [incremental model strategy](https://docs.getdbt.com/docs/build/incremental-models-overview#:~:text=Incremental%20models%20in%20dbt%20is,data%20since%20the%20last%20run.) so if we have multiple CTEs in a file, we try to break them up into their own files.

This code should also have some basic data quality checks, [unit testing](https://docs.getdbt.com/docs/build/unit-tests) so we can validated the correctness and [snapshots](https://docs.getdbt.com/docs/build/snapshots) so we can track changes over time.

File: models/marts/fueled_events_mart_cart_viewed.sql [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+select 
+  s.session_id,
+  cv.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from cart_viewed as cv
+left join session_info as s
+  on cv.anonymous_id = s.context_anonymous_id 
+qualify row_number() over(partition by cv.event_id order by abs(UNIX_SECONDS(s.session_start)-UNIX_SECONDS(cv.original_timestamp))) = 1
```
Comment (farmanp) [inline models/marts/fueled_events_mart_cart_viewed.sql:21]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained:

```
select 
  s.session_id,
  cv.*,
  s.marketing_source as session_marketing_source,
  s.marketing_medium as session_marketing_medium
from {{ ref('fueled_events_intermediate_cart_viewed') }} as cv
left join {{ ref('fueled_events_intermediate_session_info') }} as s
  on cv.anonymous_id = s.context_anonymous_id 
qualify row_number() over(partition by cv.event_id order by abs(UNIX_SECONDS(s.session_start) - UNIX_SECONDS(cv.original_timestamp))) = 1
```

File: models/marts/fueled_events_mart_logged_in.sql [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+from {{ ref('fueled_events_intermediate_session_info') }}
+)
+
+select 
+  s.session_id,
+  li.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from logged_in as li
+left join session_info as s
```
Comment (farmanp) [inline models/marts/fueled_events_mart_logged_in.sql:19]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained (see above)

File: models/marts/fueled_events_mart_pages.sql [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+from {{ ref('fueled_events_intermediate_session_info') }}
+)
+
+select 
+  s.session_id,
+  p.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from pages as p
+left join session_info as s
```
Comment (farmanp) [inline models/marts/fueled_events_mart_pages.sql:19]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained (see above)

File: models/marts/fueled_events_mart_product_added.sql [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+from {{ ref('fueled_events_intermediate_session_info') }}
+)
+
+select 
+  s.session_id,
+  pa.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from product_added as pa  
+left join session_info as s
```
Comment (farmanp) [inline models/marts/fueled_events_mart_product_added.sql:19]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained (see above)

File: models/marts/fueled_events_mart_product_list_viewed.sql [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+from {{ ref('fueled_events_intermediate_session_info') }}
+)
+
+select 
+  s.session_id,
+  plv.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from product_list_viewed as plv
+left join session_info as s
```
Comment (farmanp) [inline models/marts/fueled_events_mart_product_list_viewed.sql:19]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained (see above)

File: models/marts/fueled_events_mart_product_removed.sql [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+from {{ ref('fueled_events_intermediate_session_info') }}
+)
+
+select 
+  s.session_id,
+  pr.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from product_removed as pr
+left join session_info as s
```
Comment (farmanp) [inline models/marts/fueled_events_mart_product_removed.sql:19]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained (see above)

File: models/marts/fueled_events_mart_product_viewed.sql [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+from {{ ref('fueled_events_intermediate_session_info') }}
+)
+
+select 
+  s.session_id,
+  pv.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from product_viewed as pv  
+left join session_info as s
```
Comment (farmanp) [inline models/marts/fueled_events_mart_product_viewed.sql:19]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained (see above)

File: models/marts/fueled_events_mart_products_searched.sql [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+from {{ ref('fueled_events_intermediate_session_info') }}
+)
+
+select 
+  s.session_id,
+  ps.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from product_searched as ps  
+left join session_info as s
```
Comment (farmanp) [inline models/marts/fueled_events_mart_products_searched.sql:19]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained (see above)

File: models/marts/fueled_events_mart_signed_up.sql [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+from {{ ref('fueled_events_intermediate_session_info') }}
+)
+
+select 
+  s.session_id,
+  su.*,
+  s.marketing_source as session_marketing_soure,
+  s.marketing_medium as session_marketing_medium
+from signed_up as su
+left join session_info as s
```
Comment (farmanp) [inline models/marts/fueled_events_mart_signed_up.sql:19]: You can combine the CTEs into a single query to reduce code size, assuming readability is maintained (see above)

**Review by farmanp — COMMENTED** [review]
@shahriarkasib Thank you so much for your hard work here! Many of the comments I'm making relate to our best practices in using dbt so if you have any questions about it, feel free to ping in the channel. I'd also like Sean to see this as well so we can make sure that the goal is fulfilled for what this accomplishing. 

Some other questions I have: 
1. Are the tables available to see this transformed data? 
2. Can you add [yml documentation](https://docs.getdbt.com/reference/resource-configs/schema) for the models and include relevant data quality checks? We're introducing a lot of logic here and want to make sure it is documented.

File: models/base/scratch/fueled_events_base_session_info.sql [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  min(timestamp) over(partition by context_anonymous_id, session_id) as session_start,
+  max(timestamp) over(partition by context_anonymous_id, session_id) as session_end
+from sessions  
+order by context_anonymous_id, timestamp 
+) 
+
+select 
+  * except(landing_page_url),
+  case when landing_page_url like '%checkout%' then 'https://shop.cutsclothing.com/checkouts/'
+  when landing_page_url like '%account%' then 'https://www.cutsclothing.com/account/'
```
Comment (farmanp) [inline models/base/scratch/fueled_events_base_session_info.sql:50]: To make this code more flexible and reusable, you can replace the hardcoded URLs with variables. Here is how you can adjust the code:

```
case
  when landing_page_url like '%checkout%' then {{ var("landing_page_checkout") }}
  when landing_page_url like '%account%' then {{ var("landing_page_account") }}
  else landing_page_url
end as landing_page_url
```

And we would ensure that the variables `landing_page_checkout` and `landing_page_account` are defined in your dbt_project.yml or as environment variables.

So `dbt_project.yml`: 

```
vars:
  landing_page_checkout: 'https://shop.cutsclothing.com/checkouts/'
  landing_page_account: 'https://www.cutsclothing.com/account/'
```

This approach makes it easier to update URLs and reuse the logic across different environments or projects.

File: models/base/scratch/bigquery/fueled_events_base_product_list_viewed_products.sql [hunk 12]
```diff
@@ -1,6 +1,6 @@
 with
     unnested_product_list_viewed_products as (
-        select "product_list_viewed" as event_type, id, list_id, original_timestamp, unnested_product_column
```
[the remark this answers is not in the document]
Comment (jasoncuts) [reply models/base/scratch/bigquery/fueled_events_base_product_list_viewed_products.sql:3]: this column is not present in fueled_events table

File: models/base/scratch/fueled_events_base_product_added.sql [hunk 13]
```diff
@@ -6,7 +6,7 @@ with
         select
             brand,
             category,
-            image_url,
+            image,
             name,
```
[the remark this answers is not in the document]
Comment (jasoncuts) [reply models/base/scratch/fueled_events_base_product_added.sql:10]: this column is not present in fueled_events table

**Review by farmanp — APPROVED** [review]

File: models/base/scratch/base_scratch.yml [hunk 14]
```diff
@@ -2271,3 +2271,55 @@ models:
...
+        description: '{{ doc("col_title") }}'
+
+      - name: timestamp
+        description: 'Timestamp of when a page was viewed in a session'
+
+      - name: new_session_flag
+        description: 'Flag to identify the start of a new session'
+
+      - name: session_id
+        description: 'ID of a particular session'
```
Comment (farmanp) [inline models/base/scratch/base_scratch.yml:2313]: can we make sure to be specific about what type of session this is? "Particular" could be more descriptive here

**Pull request merged**